### ISSUE TITLE
Replace text parser with templated strings

### ISSUE THREAD

**Issue by ElAcechador** [AlderArts/foe#457]
First of all, welcome back Alder and *did you actually just convert the entire codebase to use modules instead of a labyrinth of script tags and then just port everything to Typescript like an absolute madman??* *Because that's __so fucking cool__!!*

Seriously, I've been dreaming about being able to shake a magic wand and make something like this since I first started looking at the code for this, but you actually went ahead and did it and now it's no longer a dream! But with that out of the way, this actually opens up a number of interesting things, such as now being able to use all of the newer JS features that older browsers didn't support, and I think one of those would be pretty interesting for a fext-based game like FoE.

Right now, most text containing variables is formatted roughly like this :

```ts
let parse: any = {
	playername : player.name,
	name   : kiakai.name,
};
parse = kiakai.ParserPronouns(parse);

Text.Add("You insist that [heshe] can lean on your shoulder if [heshe] needs help. [name] smiles at you, before changing the subject.", parse);
```

[Typescript has had support for ES2015 templated strings since version 1.4][1], which allows for string interpolation inside those new templated string literals, so it could now be changed to look more like this :

```ts
Text.Add(`You insist that ${kiakai.heshe()} can lean on your shoulder if ${kiakai.heshe()} needs help. ${kiakai.name} smiles at you, before changing the subject.`);
```

Or alternatively :

```ts
let heshe = kiakai.heshe();
let name = kiakai.name;

Text.Add(`You insist that ${heshe} can lean on your shoulder if ${heshe} needs help. ${name} smiles at you, before changing the subject.`);
```

This would provide some extra safety against things such as typos (since the interpreter can now validate these) without really affecting the ergonomics. It also wouldn't add much code bloat for the compiled version, it simply compiles down to string concatenation for older versions of JS. It doesn't do much in the way of type safety, however, since anything with a `toString` function is automatically converted to a string (though TSLint has [a `strict-string-expressions` option][2], which might fix that).

[1]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-1-4.html#template-strings
[2]: https://palantir.github.io/tslint/rules/strict-string-expressions/

**Comment by aldergames** [comment]
This post made my day :)

Interestingly enough, you have it completely backwards; the latest overhaul came around something like this:

- I wanna start working on the game again, but the codebase is wonky... lets start with something small.
- These template literals look kinda neat, I could maybe set up a cleaner parser with those.
- Crap... the gulp uglifier doesn't appear to support those features...
- ...I never liked gulp anyways. WebPack is kinda nice though, let's switch to that, can't be too hard...
- ...OK, so looks like I need to use require/modules in order to do this. Well, not about to stop now!
- **Silent screaming about circular dependencies**
- **Audible screaming about missing imports not showing up until the line of code actually runs**
- ...So static typing is pretty nice. I know how to set up TypeScript in WebPack, let's do that.
- **Several weeks of bashing the codebase until it stayed down**
- Hey, actually looks halfway decent now.

So in short, the entire rewrite came about from a small annoyance, and me not knowing when to quit. In the end, I feel it was worth it though, as the codebase is a lot more robust now than before.

As for the actual topic; I've given some thought to setting up a new parser system. I'll defo use template literals for all future content anyways, if nothing else than to get multiline strings so I can get rid of all the Text.NL() calls. I'm in no great hurry about it though; I'd want to set it up right instead of rushing it. I'll defo be taking some pointers from the CoC2 parser (which does use template literals and tagged templates).

**Comment by ElAcechador** [comment]
> This post made my day :)

Figured I'd let you know that your efforts definitely did not go unnoticed. ;)

>  * Crap... the gulp uglifier doesn't appear to support those features...
> 
>  * ...I never liked gulp anyways. WebPack is kinda nice though, let's switch to that, can't be too hard...

Yeah, same (even though #415 is on me, but it seemed like the best middle ground at the time, considering the amount of work needed to not just to turn this load order mess into what was then a "brand new hip JS thing" but also to keep up with upstream changes).

>  * ...OK, so looks like I need to use require/modules in order to do this. Well, not about to stop now!
> 
>  * **Silent screaming about circular dependencies**
> 
>  * **Audible screaming about missing imports not showing up until the line of code actually runs**
> 
>  * ...So static typing is pretty nice. I know how to set up TypeScript in WebPack, let's do that.
> 
>  * **Several weeks of bashing the codebase until it stayed down**
> 
>  * Hey, actually looks halfway decent now.
> 
> So in short, the entire rewrite came about from a small annoyance, and me not knowing when to quit. In the end, I feel it was worth it though, as the codebase is a lot more robust now than before.

Ah, so it was actually one of those "For want of a nail, the entire project was converted to Typescript" situation? Yeah, I can definitely relate.

> As for the actual topic; I've given some thought to setting up a new parser system. I'll defo use template literals for all future content anyways, if nothing else than to get multiline strings so I can get rid of all the Text.NL() calls. I'm in no great hurry about it though; I'd want to set it up right instead of rushing it. I'll defo be taking some pointers from the CoC2 parser (which does use template literals and tagged templates).

I wasn't aware the CoC2 codebase had been made public yet, I'll have to check that out. And yeah, I'm sure there's plenty of other of things you also want to do now that you've got your hands back into this thing. I just happenned to run into one or two text parser errors while checking out the new version (the forum seems to have no shortage of these), so that seemed like a good thing to put on the bucket list.

**Comment by aldergames** [comment]
> Ah, so it was actually one of those "For want of a nail, the entire project was converted to Typescript" situation? Yeah, I can definitely relate.

Something like that. I do feel that the TypeScript thing is worth it in and of itself regardless, due to how much more robust it makes things. That said, I recognize there were several points during the process that other decisions could have been made that solved the actual problem much quicker.

> I wasn't aware the CoC2 codebase had been made public yet, I'll have to check that out.

It's not, to my knowledge, but I've seen enough snippets of it in Fen workchat to have an idea of how it works.

On the surface, I have no issue in the examples above:
```
Text.Add(`You insist that ${kiakai.heshe()} can lean on your shoulder if ${kiakai.heshe()} needs help. ${kiakai.name} smiles at you, before changing the subject.`);
```
Usually, I use convenience functions for some of this stuff though (ParserPronouns etc) and some of those are non-trivial. So in that case in order to get proper type hinting I'd need to give a type to the return value of said functions. I'm not sure how one can properly concatenate types... but then again with this approach the functions might not be needed.

My main issue is that some of the code function names can be quite wordy, and I don't really want to clutter things up with that. So things like:
```
Text.Add(`This is a cluttered code tag: ${kiakai.FirstBreastRow().Short()}.`);
```
rather than the current:
```
Text.Add(`This is a nice code tag: [kbreasts].`, parse);
```

**Comment by aldergames** [comment]
@ElAcechador do you think you could have a look at this PR? #458

Would you have any comments or suggestions in regards to my approach?

**Comment by ElAcechador** [comment]
I was imagining using the built-in variable interpolation, but I can see why you didn't. Actually, considering the main benefit you were going for is preserving line-breaks, it makes sense that you did it like this instead.

The basic variable expansion case looks the same as what's currently there, which is fine, though I'm having trouble understanding how boolean variables and recursive interpolation are supposed to work or look like concretely. A few examples (or maybe even unit tests? for a core component like this, it'd probably be a good idea) would be useful.

**Comment by aldergames** [comment]
I can understand without examples the solution looks a bit confusing. The intended usage would be something like this:

```
enum Flag {
  Val1 = 1,
  Val2 = 2,
}
Text.Add(P2`This shows how to use conditional parsing: [if ${flag === Flag.Val1}|${flag === Val2}:
  This text is written if flag === Flag.Val1 (the variable interpolation will replace it with "true")|
  This text is written if flag === Flag.Val2, and here is and example of nested parsing: [if ${flag2}: a|b].|
  This is the else case, if flag is some other value.]
  The top level text continues here.`);
```

**Comment by ElAcechador** [comment]
> I can understand without examples the solution looks a bit confusing. The intended usage would be something like this:
> 
> ```
> enum Flag {
>   Val1 = 1,
>   Val2 = 2,
> }
> Text.Add(P2`This shows how to use conditional parsing: [if ${flag === Flag.Val1}|${flag === Val2}:
>   This text is written if flag === Flag.Val1 (the variable interpolation will replace it with "true")|
>   This text is written if flag === Flag.Val2, and here is and example of nested parsing: [if ${flag2}: a|b].|
>   This is the else case, if flag is some other value.]
>   The top level text continues here.`);
> ```

Geez, are you sure it's a good idea to cram all that inside a single run of your text parser? I know the same thing is possible with string template litterals, but at least those can be validated as correct during compilation.

> I like the idea of using the variable interpolation, that was my first approach but I didn't understand how to implement it without introducing a lot of clutter. Example above, the `${kiakai}${"heshe"}`. I would like this to at least be `${kiakai}.heshe` (which may be possible by parsing the beginning of the next literal).

I figured using interpolation for things like `` `${kiakai}${'heshe'}` `` would be beneficial given how ${'heshe'} can be validated as a one of the keys for a given class of elements (hence `...tokens: Array<keyof parseChar|Entity>`, though `parseChar` wouldn't be valid for this since it's an object and not a class, but you get the idea), so there's no risk of making a typo and having it not be caught by the compiler.

> Additionally, in this case `kiakai` has to be a locally accessible variable, which may or may not be the case. I've added these throughout a lot of functions already for the old parser implementation, but in my new implementation, they are not actually needed (a very small gain, granted).

I haven't looked at the code in detail for a while now, but IIRC the character objects for a given scene are pretty much always locally available, aren't they? That's how I remember it, anyway, given how the parser object needs to be setup with the variable substitutions.

> My current code is losely based on ideas from this article and other similar things around the Internet: https://jaysoo.ca/2014/03/20/i18n-with-es2015-template-literals/

The mix of native string template interpolation and surrounding string parsing looks like something that would drive me just a little bit mad, but I can see what you're going for, yeah.

**Comment by aldergames** [comment]
You bring up valid points. What I'd ideally want is a solution that is compile-time validated (both my current and new parser fails at this), and that is also terse and readable in its presentation (in my opinion, the pure template literal approach fails in this).

> Geez, are you sure it's a good idea to cram all that inside a single run of your text parser?

Sure? By no means.

> I figured using interpolation for things like `${kiakai}${'heshe'}` would be beneficial given how ${'heshe'} can be validated as a one of the keys for a given class of elements (hence ...tokens: Array<keyof parseChar|Entity>, though parseChar wouldn't be valid for this since it's an object and not a class, but you get the idea), so there's no risk of making a typo and having it not be caught by the compiler.

That's clever! I think an enum could be used to accomplish that.

> I haven't looked at the code in detail for a while now, but IIRC the character objects for a given scene are pretty much always locally available, aren't they? That's how I remember it, anyway, given how the parser object needs to be setup with the variable substitutions.

Correct, this is mostly the case at the moment. What I'm saying it wouldn't necessarily have to be with the new approach, since it would fetch the objects inside the parser itself through the GAME object.

In the end, I will sacrifice the validation for a reduction in time from writing to code (which is a significant amount for complex branching scenes with custom logic). Maybe at least in part I can circumvent the issue by creating a linter of sorts, which shouldn't be that much work.

**Comment by ElAcechador** [comment]
>

**Comment by ElAcechador** [comment]
> You bring up valid points. What I'd ideally want is a solution that is compile-time validated (both my current and new parser fails at this), and that is also terse and readable in its presentation (in my opinion, the pure template literal approach fails in this).

> In the end, I will sacrifice the validation for a reduction in time from writing to code (which is a significant amount for complex branching scenes with custom logic). Maybe at least in part I can circumvent the issue by creating a linter of sorts, which shouldn't be that much work.

The only terseness and readability issue I'm seeing with string templates is that variables have to be in-scope and the tokens being interpolated have to be valid JS expressions, which your parser seems to address to a certain degree, but (at least, going with the examples you've given thus far) not to the degree where I'd say it's worth tossing runtime validation aside. Letting the parser deal with the entity objects on its own could make sense (although wouldn't that mean hardcoding all characters like kia, momo, terry and so on in at least one section of the game code?), though if it were up to me, I'd definitely prefer something like having all interpolable values come from an enum like yiu mentioned and not do any of that substitution by hand, but that's just me.

Otherwise, in terms of features, it seems to support everything your custom syntax would (including nesting, since you can nest template literals inside other template literals, though I don't know how well those would interact with enums). The one thing I would see be an eventual problem us inline branching, since while JS has ternary expressions to let you pick between two branches, having an else case like in your example above would be trickier.

That's how I see it, anyway, from the vewpoint of a guy who has a history of filing the edges of square pegs to make sure they fit nice and snug into round holes. If you say that way of doing it is going to make text-to-code conversion easier in the long run, I figure you'd certainly know better than me.

**Comment by ElAcechador** [comment]
Dropping the tagged templates for generic ones, I see? Makes sense, even though there are some benefits to be had from tagged templates (namely type restriction, since string templates accept anything that can be converted to a string, which includes literally everything by default), but this is probably going to be easier to maintain, and the primary argument for type safety had more to do with typos and missing variables than with incorrect types.

This is pretty much the sort of thing I had in mind initially (though using getters instead of functions goes a long way in improving the ergonomics), and it does look pretty clean. Mad props to you, man!

**Comment by aldergames** [comment]
> (namely type restriction, since string templates accept anything that can be converted to a string, which includes literally everything by default)

I've disabled implicit string conversion in my tslint pass, so this should not be an issue either.

**Comment by ElAcechador** [comment]
Yeah, the only "fix" I could find that worked in that case would be to give everything a null prototype or to remove the toString method on the object prototype, but both would be pretty drastic and I'm not sure Typescript or TSLint would even recognize that. But like I said, that's hardly the primary concern (unless it is, but that would be pretty hard to fix without being overly restrictive in a number of other ways) and the current solution should be good enough. I doubt you're likely to go ``Test.Add(`You insist that ${kiakai}...`)`` and forget the actual getter very often.

**Comment by aldergames** [comment]
Thank you for your investigation! Like you say, the main issue here isn't type strictness but wether the attribute exists at all. Using the EntityParser should alleviate that somewhat, even if there's no enforcement of actually having to use it. I think I'll clean up the current code and merge it in the near future, removing the tagged parser attempt, and then I'll close this issue.

**Comment by aldergames** [comment]
Merged #458
